File: python-service/src/string_service.py
"""
String Service for pytest unit testing
"""

import re
from typing import List, Optional


class StringService:
    """Service class for string operations"""
    
    def reverse(self, text: Optional[str]) -> Optional[str]:
        """Reverse a string"""
        if text is None:
            return None
        return text[::-1]
    
    def is_palindrome(self, text: Optional[str]) -> bool:
        """Check if a string is palindrome"""
        if text is None:
            return False
        # Remove non-alphanumeric characters and convert to lowercase
        cleaned = re.sub(r'[^a-zA-Z0-9]', '', text.lower())
        return cleaned == cleaned[::-1]
    
    def count_vowels(self, text: Optional[str]) -> int:
        """Count vowels in a string"""
        if text is None:
            return 0
        vowels = 'aeiouAEIOU'
        return sum(1 for char in text if char in vowels)
    
    def to_title_case(self, text: Optional[str]) -> Optional[str]:
        """Convert string to title case"""
        if text is None or text.strip() == "":
            return text
        return text.title()
    
    def remove_duplicates(self, text: Optional[str]) -> Optional[str]:
        """Remove duplicate characters from string"""
        if text is None:
            return None
        seen = set()
        result = ""
        for char in text:
            if char not in seen:
                seen.add(char)
                result += char
        return result
    
    def find_longest_word(self, text: Optional[str]) -> str:
        """Find the longest word in a string"""
        if text is None or text.strip() == "":
            return ""
        words = text.split()
        if not words:
            return ""
        return max(words, key=len)
    
    def count_words(self, text: Optional[str]) -> int:
        """Count words in a string"""
        if text is None or text.strip() == "":
            return 0
        return len(text.split())
    
    def is_anagram(self, text1: Optional[str], text2: Optional[str]) -> bool:
        """Check if two strings are anagrams"""
        if text1 is None or text2 is None:
            return False
        # Remove spaces and convert to lowercase
        text1_clean = re.sub(r'\s', '', text1.lower())
        text2_clean = re.sub(r'\s', '', text2.lower())
        return sorted(text1_clean) == sorted(text2_clean)
    
    def capitalize_words(self, text: Optional[str]) -> Optional[str]:
        """Capitalize first letter of each word"""
        if text is None:
            return None
        return ' '.join(word.capitalize() for word in text.split())
    
    def count_occurrences(self, text: Optional[str], substring: str) -> int:
        """Count occurrences of a substring in a string"""
        if text is None or substring is None:
            return 0
        return text.count(substring)
    
    def is_valid_email(self, email: Optional[str]) -> bool:
        """Check if email is valid"""
        if email is None:
            return False
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return bool(re.match(pattern, email))
    
    def extract_numbers(self, text: Optional[str]) -> List[int]:
        """Extract all numbers from a string"""
        if text is None:
            return []
        numbers = re.findall(r'\d+', text)
        return [int(num) for num in numbers]
    
    def remove_special_chars(self, text: Optional[str]) -> Optional[str]:
        """Remove special characters from string"""
        if text is None:
            return None
        # Replace special characters with spaces, then clean up
        result = re.sub(r'[^a-zA-Z0-9\s]', ' ', text)
        # Replace multiple spaces with single space
        result = re.sub(r'\s+', ' ', result)
        return result.strip()

File: java-service/src/main/java/com/example/StringService.java
package com.example;

import org.springframework.stereotype.Service;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class StringService {
    
    /**
     * Reverses a string
     */
    public String reverse(String input) {
        if (input == null) {
            return null;
        }
        return new StringBuilder(input).reverse().toString();
    }
    
    /**
     * Checks if a string is palindrome
     */
    public boolean isPalindrome(String input) {
        if (input == null) {
            return false;
        }
        String cleaned = input.toLowerCase().replaceAll("[^a-zA-Z0-9]", "");
        return cleaned.equals(reverse(cleaned));
    }
    
    /**
     * Counts vowels in a string
     */
    public int countVowels(String input) {
        if (input == null) {
            return 0;
        }
        return (int) input.toLowerCase().chars()
                .mapToObj(ch -> (char) ch)
                .filter(ch -> "aeiou".indexOf(ch) != -1)
                .count();
    }
    
    /**
     * Converts string to title case
     */
    public String toTitleCase(String input) {
        if (input == null || input.trim().isEmpty()) {
            return input;
        }
        
        String[] words = input.toLowerCase().split("\\s+");
        List<String> titleCaseWords = Arrays.stream(words)
                .map(word -> word.isEmpty() ? word : 
                    Character.toUpperCase(word.charAt(0)) + word.substring(1))
                .collect(Collectors.toList());
        
        return String.join(" ", titleCaseWords);
    }
    
    /**
     * Removes duplicate characters from string
     */
    public String removeDuplicates(String input) {
        if (input == null) {
            return null;
        }
        return input.chars()
                .distinct()
                .mapToObj(ch -> String.valueOf((char) ch))
                .collect(Collectors.joining());
    }
    
    /**
     * Finds the longest word in a string
     */
    public String findLongestWord(String input) {
        if (input == null || input.trim().isEmpty()) {
            return "";
        }
        
        return Arrays.stream(input.split("\\s+"))
                .max((a, b) -> Integer.compare(a.length(), b.length()))
                .orElse("");
    }
    
    /**
     * Counts words in a string
     */
    public int countWords(String input) {
        if (input == null || input.trim().isEmpty()) {
            return 0;
        }
        return input.trim().split("\\s+").length;
    }
    
    /**
     * Checks if two strings are anagrams
     */
    public boolean isAnagram(String text1, String text2) {
        if (text1 == null || text2 == null) {
            return false;
        }
        String cleaned1 = text1.toLowerCase().replaceAll("[^a-zA-Z]", "");
        String cleaned2 = text2.toLowerCase().replaceAll("[^a-zA-Z]", "");
        
        if (cleaned1.length() != cleaned2.length()) {
            return false;
        }
        
        char[] chars1 = cleaned1.toCharArray();
        char[] chars2 = cleaned2.toCharArray();
        Arrays.sort(chars1);
        Arrays.sort(chars2);
        
        return Arrays.equals(chars1, chars2);
    }
}
